pkg/scheduler/framework/v1alpha1/waiting_pods_map.go

```diff
@@ -0,0 +1,116 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"sync"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// waitingPodsMap a thread-safe map used to maintain pods waiting in the permit phase.
+type waitingPodsMap struct {
+	pods map[types.UID]WaitingPod
+	mu   sync.RWMutex
+}
+
+// NewWaitingPodsMap returns a new waitingPodsMap.
+func NewWaitingPodsMap() *waitingPodsMap {
+	return &waitingPodsMap{
+		pods: make(map[types.UID]WaitingPod),
+	}
+}
+
+// Add a new WaitingPod to the map.
+func (m *waitingPodsMap) Add(uid types.UID, wp WaitingPod) {
```

Comment:
I think we should make "Add" and "Remove" private. Only the framework should be able to add or remove waiting pods. Also, if these functions are called by a "callback" function passed to "Iterate", they will cause a deadlock.